src/server/services/wms/qgswmsgetlegendgraphics.cpp

```diff
@@ -50,42 +53,60 @@ namespace QgsWms
     context.setFlag( QgsWmsRenderContext::UseSrcWidthHeight );
     context.setParameters( parameters );
 
-    const QString format = request.parameters().value( QStringLiteral( "FORMAT" ), QStringLiteral( "PNG" ) );
-    ImageOutputFormat outputFormat = parseImageFormat( format );
+    // get the requested output format
+    QgsWmsParameters::Format format = parameters.format();
 
-    QString saveFormat;
-    QString contentType;
-    switch ( outputFormat )
+    // parameters.format() returns NONE if the requested format is image/png with a
+    // mode (e.g. image/png;mode=16bit), so in that case we use parseImageFormat to
+    // give the requested format another chance
+
+    QString imageSaveFormat;
+    QString imageContentType;
+    if ( format == QgsWmsParameters::Format::PNG )
+    {
+      imageContentType = "image/png";
+      imageSaveFormat = "PNG";
+    }
+    else if ( format == QgsWmsParameters::Format::JPG )
+    {
+      imageContentType = "image/jpeg";
+      imageSaveFormat = "JPEG";
+    }
+    else if ( format == QgsWmsParameters::Format::NONE )
+    {
+      switch ( parseImageFormat( parameters.formatAsString() ) )
+      {
+        case PNG:
+        case PNG8:
+        case PNG16:
+        case PNG1:
+          format = QgsWmsParameters::Format::PNG;
+          imageContentType = "image/png";
+          imageSaveFormat = "PNG";
+          break;
+        default:
+          break;
+      }
+    }
+
+    if ( format == QgsWmsParameters::Format::NONE )
     {
-      case PNG:
-      case PNG8:
-      case PNG16:
-      case PNG1:
-        contentType = "image/png";
-        saveFormat = "PNG";
-        break;
-      case JPEG:
-        contentType = "image/jpeg";
-        saveFormat = "JPEG";
-        break;
-      default:
-        throw QgsServiceException( "InvalidFormat",
-                                   QStringLiteral( "Output format '%1' is not supported in the GetLegendGraphic request" ).arg( format ) );
-        break;
+      throw QgsServiceException( "InvalidFormat",
+                                 QStringLiteral( "Output format '%1' is not supported in the GetLegendGraphic request" ).arg( parameters.formatAsString() ) );
```

Comment:
Can you use the `QgsBadRequestException` class with the `QgsServiceException::OGC_InvalidFormat` enum instead please?

Reply:
This is code that I just moved around. So I didn't want to change it. But I followed your suggestion and added a commit to change that.